### PR TITLE
LibWeb/CSS: Check if percentage has a value

### LINKED ISSUE #3050: Crash on open.spotify.com for CSSNumericType::apply_percent_hint
### Summary

Spotify crashes here: https://github.com/LadybirdBrowser/ladybird/blob/062e33438edd7108ce3b446cea85bb6c0270e89b/Libraries/LibWeb/CSS/CSSNumericType.cpp#L304

Invoked from: https://github.com/LadybirdBrowser/ladybird/blob/master/Libraries/LibWeb/CSS/CSSNumericType.cpp#L167

When applying percent hint.

### Operating system

Linux

### Steps to reproduce

1. Open open.spotify.com in Ladybird
2.🔥

### Expected behavior

Doesn't crash.

### Actual behavior

Crashes.

### URL for a reduced test case

open.spotify.com

### HTML/SVG/etc. source for a reduced test case

```html
Needs reduction
```

### Log output and (if possible) backtrace

```shell
VERIFICATION FAILED: !percent_hint().has_value() at /home/shannon/personal/ladybird/Libraries/LibWeb/CSS/CSSNumericType.cpp:311
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-ak.so.0(ak_trap+0x35) [0x7f4b94f851a5]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-ak.so.0(+0x154cf) [0x7f4b94f854cf]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::CSSNumericType::apply_percent_hint(Web::CSS::CSSNumericType::BaseType) 0x8a) [0x7f4b9609a1aa]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::CSSNumericType::added_to(Web::CSS::CSSNumericType const&) const 0x3e5) [0x7f4b96099d95]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0(+0x86ecb1) [0x7f4b961a7cb1]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::SumCalculationNode::create(AK::Vector<AK::NonnullOwnPtr<Web::CSS::CalculationNode>, 0ul>) 0x2d) [0x7f4b961a7a9d]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::convert_to_calculation_node(AK::Variant<Web::CSS::Parser::CalcParsing::Operator, Web::CSS::Number, Web::CSS::Parser::Dimension, Web::CSS::CalculationNode::ConstantType, AK::NonnullOwnPtr<Web::CSS::Parser::CalcParsing::ProductNode>, AK::NonnullOwnPtr<Web::CSS::Parser::CalcParsing::SumNode>, AK::NonnullOwnPtr<Web::CSS::Parser::CalcParsing::InvertNode>, AK::NonnullOwnPtr<Web::CSS::Parser::CalcParsing::NegateNode>, AK::NonnullRawPtr<Web::CSS::Parser::ComponentValue const> > const&) 0x9b1) [0x7f4b96135ae1]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::parse_a_calculation(AK::Vector<Web::CSS::Parser::ComponentValue, 0ul> const&) 0x1480) [0x7f4b960fa050]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::parse_calculated_value(Web::CSS::Parser::ComponentValue const&) 0x5b) [0x7f4b960f89db]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::parse_css_value_for_properties(AK::Span<Web::CSS::PropertyID const>, Web::CSS::Parser::TokenStream<Web::CSS::Parser::ComponentValue>&) 0x1b2a) [0x7f4b9610b5fa]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::parse_css_value_for_property(Web::CSS::PropertyID, Web::CSS::Parser::TokenStream<Web::CSS::Parser::ComponentValue>&) 0x31) [0x7f4b96107291]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::parse_css_value(Web::CSS::PropertyID, Web::CSS::Parser::TokenStream<Web::CSS::Parser::ComponentValue>&, AK::Optional<AK::String>) 0xe19) [0x7f4b960f7289]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::resolve_unresolved_style_value(Web::DOM::Element&, AK::Optional<Web::CSS::Selector::PseudoElement::Type>, Web::CSS::PropertyID, Web::CSS::UnresolvedStyleValue const&) 0x22c) [0x7f4b9613644c]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::Parser::Parser::resolve_unresolved_style_value(Web::CSS::Parser::ParsingContext const&, Web::DOM::Element&, AK::Optional<Web::CSS::Selector::PseudoElement::Type>, Web::CSS::PropertyID, Web::CSS::UnresolvedStyleValue const&) 0x82) [0x7f4b961361d2]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::StyleComputer::cascade_declarations(Web::CSS::CascadedProperties&, Web::DOM::Element&, AK::Optional<Web::CSS::Selector::PseudoElement::Type>, AK::Vector<Web::CSS::MatchingRule, 0ul> const&, Web::CSS::CascadeOrigin, Web::CSS::Important, AK::Optional<AK::FlyString>) const 0x262) [0x7f4b96189a82]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::StyleComputer::compute_cascaded_values(Web::DOM::Element&, AK::Optional<Web::CSS::Selector::PseudoElement::Type>, bool&, Web::CSS::StyleComputer::ComputeStyleMode) const 0xe49) [0x7f4b9618c559]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::StyleComputer::compute_style_impl(Web::DOM::Element&, AK::Optional<Web::CSS::Selector::PseudoElement::Type>, Web::CSS::StyleComputer::ComputeStyleMode) const 0x7c) [0x7f4b96190e2c]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::CSS::StyleComputer::compute_style(Web::DOM::Element&, AK::Optional<Web::CSS::Selector::PseudoElement::Type>) const 0x35) [0x7f4b96190d85]
/home/shannon/personal/ladybird/Build/debug/lib/liblagom-web.so.0 Web::DOM::Element::recompute_style() 0x66) [0x7f4b96228056]
```

### Screenshots or screen recordings

_No response_

### Build flags or config settings

_No response_

### Contribute a patch?

- [ ] I’ll contribute a patch for this myself.

### PR DESCRIPTION
**Pull request by ChaseKnowlden**
Fixes crash on open.spotify.com when loading the page.

Fixes #3050

### CODE REVIEW

File: Libraries/LibWeb/CSS/CSSNumericType.cpp [hunk 1]
```diff
@@ -145,16 +145,16 @@ Optional<CSSNumericType> CSSNumericType::added_to(CSSNumericType const& other) c
...
+        // finalType doesn't already contain. Set finalType's percent hint to type1's percent hint. Return finalType.
         final_type.copy_all_entries_from(type1, SkipIfAlreadyPresent::No);
         final_type.copy_all_entries_from(type2, SkipIfAlreadyPresent::Yes);
         final_type.set_percent_hint(type1.percent_hint());
         return final_type;
     }
     //    If type1 and/or type2 contain "percent" with a non-zero value,
     //    and type1 and/or type2 contain a key other than "percent" with a non-zero value
-    if ((type1.exponent(BaseType::Percent) != 0 || type2.exponent(BaseType::Percent) != 0)
+    if ((type1.exponent(BaseType::Percent).has_value() || type2.exponent(BaseType::Percent).has_value())
```
Comment (LucasChollet) [inline Libraries/LibWeb/CSS/CSSNumericType.cpp:157]: The spec says it needs to be a non-zero value, I think the old code was valid.

Comment (shannonbooth) [reply Libraries/LibWeb/CSS/CSSNumericType.cpp:157]: Perhaps the language of the spec is a bit subtle here. Maybe by value, it means non-null. In which case this should still be checking for has_value, as well as for non-zero.

i.e

percent1.has_value() && percent1.value() != 0

Comment (LucasChollet) [reply Libraries/LibWeb/CSS/CSSNumericType.cpp:157]: > percent1.has_value() && percent1.value() != 0

Isn't that exactly what `percent1 != 0` does?

Comment (shannonbooth) [reply Libraries/LibWeb/CSS/CSSNumericType.cpp:157]: null is not equal to zero, so i dont believe so

Comment (LucasChollet) [reply Libraries/LibWeb/CSS/CSSNumericType.cpp:157]: Oh ok, my bad, I'm just understanding what you're saying.

**Comment by ChaseKnowlden** [comment]
Closing then.

**Pull request closed without merging**